beacon-chain/core/validators/validator.go

```diff
@@ -310,105 +316,70 @@ func SlashValidator(state *pb.BeaconState, idx uint64) (*pb.BeaconState, error)
-//    max_balance_churn = max(
-//        MAX_DEPOSIT_AMOUNT,
-//        total_balance // (2 * MAX_BALANCE_CHURN_QUOTIENT)
-//    )
-//
-//    # Activate validators within the allowable balance churn
-//    balance_churn = 0
-//    for index, validator in enumerate(state.validator_registry):
-//        if validator.activation_epoch > get_entry_exit_effect_epoch(current_epoch) and state.validator_balances[index] >= MAX_DEPOSIT_AMOUNT:
-//            # Check the balance churn would be within the allowance
-//            balance_churn += get_effective_balance(state, index)
-//            if balance_churn > max_balance_churn:
-//                break
-//
-//            # Activate validator
-//            activate_validator(state, index, is_genesis=False)
-//
-//    # Exit validators within the allowable balance churn
-//    balance_churn = 0
-//    for index, validator in enumerate(state.validator_registry):
-//        if validator.exit_epoch > get_entry_exit_effect_epoch(current_epoch) and validator.status_flags & INITIATED_EXIT:
-//            # Check the balance churn would be within the allowance
-//            balance_churn += get_effective_balance(state, index)
-//            if balance_churn > max_balance_churn:
-//                break
-//
-//            # Exit validator
-//            exit_validator(state, index)
-//
-//    state.validator_registry_update_epoch = current_epoch
-func UpdateRegistry(state *pb.BeaconState) (*pb.BeaconState, error) {
+//   def process_registry_updates(state: BeaconState) -> None:
+//     # Process activation eligibility and ejections
+//     for index, validator in enumerate(state.validator_registry):
+//         if validator.activation_eligibility_epoch == FAR_FUTURE_EPOCH and validator.effective_balance >= MAX_EFFECTIVE_BALANCE:
+//             validator.activation_eligibility_epoch = get_current_epoch(state)
+//         if is_active_validator(validator, get_current_epoch(state)) and validator.effective_balance <= EJECTION_BALANCE:
+//             initiate_validator_exit(state, index)
+//     # Queue validators eligible for activation and not dequeued for activation prior to finalized epoch
+//     activation_queue = sorted([
+//         index for index, validator in enumerate(state.validator_registry) if
+//         validator.activation_eligibility_epoch != FAR_FUTURE_EPOCH and
+//         validator.activation_epoch >= get_delayed_activation_exit_epoch(state.finalized_epoch)
+//     ], key=lambda index: state.validator_registry[index].activation_eligibility_epoch)
+//     # Dequeued validators for activation up to churn limit (without resetting activation epoch)
+//     for index in activation_queue[:get_churn_limit(state)]:
+//         validator = state.validator_registry[index]
+//         if validator.activation_epoch == FAR_FUTURE_EPOCH:
+//             validator.activation_epoch = get_delayed_activation_exit_epoch(get_current_epoch(state))
+func UpdateRegistry(state *pb.BeaconState) *pb.BeaconState {
```

Comment:
On a second thought, can we move this under `epoch_processing.go` and rename the function to `ProcessRegistryUpdates`? Thanks,

Reply:
done